From patchwork Wed Aug 29 13:40:50 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Quan Xu <quan.xu0@gmail.com>
X-Patchwork-Id: 10580353
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7886A14BD
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 29 Aug 2018 13:40:59 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 662E52B250
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 29 Aug 2018 13:40:59 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 63BAE2B234; Wed, 29 Aug 2018 13:40:59 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 040432B1FB
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 29 Aug 2018 13:40:58 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727651AbeH2Rh5 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 29 Aug 2018 13:37:57 -0400
Received: from mail-oi0-f68.google.com ([209.85.218.68]:38345 "EHLO
        mail-oi0-f68.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727312AbeH2Rh5 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 29 Aug 2018 13:37:57 -0400
Received: by mail-oi0-f68.google.com with SMTP id x197-v6so9198758oix.5
        for <kvm@vger.kernel.org>; Wed, 29 Aug 2018 06:40:57 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=to:cc:from:subject:message-id:date:user-agent:mime-version
         :content-transfer-encoding;
        bh=3kGx/jhhRySX1xUrDiTrLHtc1yxiBqAblV73iS4pIgo=;
        b=GXyZW9YG0x8VTWn+jDi9KKlyCuCDJxCrbvk4i5UMyMSt80fHyQIeHHodxojdu1eKzC
         VRDBtfRbEgtsq9MhALDDDZ7l1x6QtasG/iTwT4KTcMIfG1l3pRecycS9jA+31XJqKt7v
         VgwIe3fE2heHkyLi9bssucO2Bnwrj0JfnrqzrtneZ5giES5Xrx0/2WW124IZPBFuSUwm
         B7EWMM9vxlcp9Mxg/svvKI6x9jAh6WV/JGl7Eh10IkkrA0OH1iCT+8euVufEZ0uezIiT
         1ST7QinDhs/VSzNHZkvixqRnDaqbFQ0XiWbif2UYE+B6q+XGGxl2fIAFUgcX7Sd5/kLE
         CPrQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:to:cc:from:subject:message-id:date:user-agent
         :mime-version:content-transfer-encoding;
        bh=3kGx/jhhRySX1xUrDiTrLHtc1yxiBqAblV73iS4pIgo=;
        b=kbkFX2QLiOiWcDzVqwnODRkw3Xx/OqCd3USZ5Oc9mzDNAYBhgkZMtjpVXSQa1l/RHv
         VsX3F4lzOM97TlqmqlFf8+zniwhdBxM5L43zJdfMb0CPo2LV3AIuFtQ4Rh7Wpx92/RL3
         jMY2GKaSyWHVZy4PptSopRiwDCnRoYV2jHgTox5mX79Xy+tAJuJ3pwnKRSmKlojDpTYN
         DYC8kSjoV+YfZIFxIrblfvMmuZaIW2AJOoHfpli1bjJRmhK8t2aWNUJcGhZIXp/0ErIh
         7BzKHfvTh40JmhCXtz3pUQDfwpc8QzZKETeYXeJ9WfqpbjDmeidDjGPSVhpD3r2drjGa
         atEw==
X-Gm-Message-State: APzg51AL+GzGSYrHK/J67OLvfQj2wUs3jiE/itn7nSWqLl5w45j44Hu1
        G1ciD5F2RKOIpEwH435pIGISdq6f
X-Google-Smtp-Source: 
 ANB0VdYVkLBU/sIt72P2FLLF/4tM7DuxVZwyRP2MBrfOrNvFPsc8dee70kIm/67E3e7eBElhWIDztw==
X-Received: by 2002:aca:b6c5:: with SMTP id
 g188-v6mr2551298oif.283.1535550056618;
        Wed, 29 Aug 2018 06:40:56 -0700 (PDT)
Received: from [30.43.105.7] ([47.89.83.49])
        by smtp.gmail.com with ESMTPSA id
 w84-v6sm3250167oie.40.2018.08.29.06.40.54
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Wed, 29 Aug 2018 06:40:56 -0700 (PDT)
To: qemu-devel@nongnu.org
Cc: quintela@redhat.com, dgilbert@redhat.com, kvm <kvm@vger.kernel.org>
From: Quan Xu <quan.xu0@gmail.com>
Subject: [PATCH RFC] migration: make sure to run iterate precopy during the
 bulk stage
Message-ID: <5ab76c3e-9310-0e08-2f1b-4ff52bf229f8@gmail.com>
Date: Wed, 29 Aug 2018 21:40:50 +0800
User-Agent: Mozilla/5.0 (Windows NT 6.1; WOW64; rv:60.0) Gecko/20100101
 Thunderbird/60.0
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8; format=flowed
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From 8dbf7370e7ea1caab0b769d0d4dcdd072d14d421 Mon Sep 17 00:00:00 2001
From: Quan Xu <quan.xu0@gmail.com>
Date: Wed, 29 Aug 2018 21:33:14 +0800
Subject: [PATCH RFC] migration: make sure to run iterate precopy during the
  bulk stage

Since the bulk stage assumes in (migration_bitmap_find_dirty) that every
page is dirty, return a rough total ram as pending size to make sure that
migration thread continues to run iterate precopy during the bulk stage.

Otherwise the downtime grows unpredictably, as migration thread needs to
send both the rest of pages and dirty pages during complete precopy.

Signed-off-by: Quan Xu <quan.xu0@gmail.com>
---
  migration/ram.c | 3 ++-
  1 file changed, 2 insertions(+), 1 deletion(-)


--
1.8.3.1

diff --git a/migration/ram.c b/migration/ram.c
index 79c8942..cfa304c 100644
--- a/migration/ram.c
+++ b/migration/ram.c
@@ -3308,7 +3308,8 @@ static void ram_save_pending(QEMUFile *f, void 
*opaque, uint64_t max_size,
          /* We can do postcopy, and all the data is postcopiable */
          *res_compatible += remaining_size;
      } else {
-        *res_precopy_only += remaining_size;
+        *res_precopy_only += (rs->ram_bulk_stage ?
+                              ram_bytes_total() : remaining_size);
      }
  }
